### ISSUE TITLE
test_delegation_refleak and test_no_leaking_notifiers from traits.tests.test_regression.TestRegression fail, at least sometimes

### ISSUE THREAD

**Issue by mwhudson** [enthought/traits#363]
They fail every time for me when I run the whole suite:

```
(master)mwhudson@aeglos:/opt/opensource/traits$ while :; do python2.7 -m unittest discover || break; done
...''
...........................ssss...............................................................................................................................................................................................................................................................................................................................................................F...F................................s..................sssssssss.......................................................................................s.s.
======================================================================
FAIL: test_delegation_refleak (traits.tests.test_regression.TestRegression)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/opensource/traits/traits/tests/test_regression.py", line 190, in test_delegation_refleak
    self.assertEqual(counts[warmup:-1], counts[warmup+1:])
AssertionError: Lists differ: [32456, 32531, 32606, 32681] != [32531, 32606, 32681, 32756]

First differing element 0:
32456
32531

- [32456, 32531, 32606, 32681]
?  -------

+ [32531, 32606, 32681, 32756]
?                     +++++++


======================================================================
FAIL: test_no_leaking_notifiers (traits.tests.test_regression.TestRegression)
Extended trait change notifications should not leaf
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/opensource/traits/traits/tests/test_regression.py", line 142, in test_no_leaking_notifiers
    self.assertEqual(len(ctrait._notifiers(1)), 0)
AssertionError: 1 != 0

----------------------------------------------------------------------
Ran 541 tests in 16.255s

FAILED (failures=2, skipped=16)
```

but pass reliably when I just run the traits.tests.test_regression.TestRegression so I suspect some kind of test isolation failure.

They also fail on the ubuntu builders, which is how I found this problem: https://launchpad.net/ubuntu/+source/python-traits/4.6.0-1build1

**Comment by mdickinson** [comment]
Thanks for the report. I'm failing to reproduce on OS X or on CentOS (with gcc 4.4.7), but we can reproduce on Ubuntu 14.04 with gcc 4.8.4.

**Comment by mdickinson** [comment]
Looks like this is a result of `test_change_except_hooks_after_exception` in `test_trait_change_event_tracer` not cleaning up after itself. I'm not sure why this issue wasn't showing up on CentOS or OS X; it could simply be down to a difference in test run order.

**Comment by mdickinson** [comment]
Fix for the broken test cleanup is in #365. Ideally, we'd fix the refleak tests to be less fragile, too: it would be better if they targeted the leak they were testing for more directly. (OTOH, this issue shows that it's useful to have a test that's detecting general reference leaks from the test suite.)

**Comment by mdickinson** [comment]
And I verified that on OS X, I wasn't seeing the error because `test_regression` was running earlier than `test_trait_change_event_tracer`.
